src/compose/compose-collection.ts

```diff
@@ -16,68 +16,132 @@ import { resolveBlockMap } from './resolve-block-map.js'
 import { resolveBlockSeq } from './resolve-block-seq.js'
 import { resolveFlowCollection } from './resolve-flow-collection.js'
 
+function resolveCollection(
+  CN: ComposeNode,
+  ctx: ComposeContext,
+  token: BlockMap | BlockSequence | FlowCollection,
+  onError: ComposeErrorHandler,
+  tagName: string | null,
+  tag?: CollectionTag
+) {
+  const coll =
+    token.type === 'block-map'
+      ? resolveBlockMap(CN, ctx, token, onError, tag)
+      : token.type === 'block-seq'
+      ? resolveBlockSeq(CN, ctx, token, onError, tag)
+      : resolveFlowCollection(CN, ctx, token, onError, tag)
+
+  const Coll = coll.constructor as typeof YAMLMap | typeof YAMLSeq
+
+  // If we got a tagName matching the class, or the tag name is '!',
+  // then use the tagName from the node class used to create it.
+  if (tagName === '!' || tagName === Coll.tagName) {
+    coll.tag = Coll.tagName
+    return coll
+  } else if (tagName) {
+    coll.tag = tagName
+  }
+
+  return coll
+}
+
 export function composeCollection(
   CN: ComposeNode,
   ctx: ComposeContext,
   token: BlockMap | BlockSequence | FlowCollection,
   tagToken: SourceToken | null,
   onError: ComposeErrorHandler
 ) {
+  const tagName: string | null = !tagToken
+    ? null
+    : ctx.directives.tagName(tagToken.source, msg =>
+        onError(tagToken, 'TAG_RESOLVE_FAILED', msg)
+      )
+
   let coll: YAMLMap.Parsed | YAMLSeq.Parsed
-  switch (token.type) {
-    case 'block-map': {
-      coll = resolveBlockMap(CN, ctx, token, onError)
-      break
-    }
-    case 'block-seq': {
-      coll = resolveBlockSeq(CN, ctx, token, onError)
-      break
-    }
-    case 'flow-collection': {
-      coll = resolveFlowCollection(CN, ctx, token, onError)
-      break
-    }
+  let expType: 'map' | 'seq' | undefined =
+    token.type === 'block-map'
+      ? 'map'
+      : token.type === 'block-seq'
+      ? 'seq'
+      : token.type === 'flow-collection'
```

Comment:
I can have it just skip the last conditional and test for `{` rather than `flow-collection`, if you're confident that no other collection token types might ever sneak in there.  I like to cover the "impossible" cases in "conventional enums" like this, because I have a tendency to add another possibility in the future and forget to update all the places it's used.

Reply:
I'm confident. `composeCollection()` is only ever called from here:
https://github.com/eemeli/yaml/blob/a67a019afe9c8f785fdfd6255b6b69d97a8bd385/src/compose/compose-node.ts#L57-L60
The parent `switch` makes sure there that `token.type` is one of the three expected, and the TS type of the argument will keep us honest should the call structure change.